### PR TITLE
Add expectedWithParameters to PrivateConstructorChecker. Fixes #16

### LINKED ISSUE #16: Private Constructor with parameter
I want to test a class that it has only exactly one constructor which is private and takes one argument. 

I thought

``` java
PrivateConstructorChecker.forClass(MyClass.class).check();
```

would do it but it gave me an exception back:

> java.lang.AssertionError: class onactivityresult.compiler.Parameter has non-default constructor with some parameters

So I guess this was a design decision. Would you be open for a new API to test the described behavior?

Maybe something alone the lines of

``` java
PrivateConstructorChecker.forClass(MyClass.class).expectWithParameters(...).check();
```

Obviously `expectWithParameters` would return a different builder so you're not able to call `expectedExceptionMessage` and `expectedTypeOfException` on it.

For the case of multiple constructors we could do `expectMultipleWithParameters(`.

### PR DESCRIPTION
**Pull request by vanniktech**
First version. Feedback is appreciated.

### CODE REVIEW

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 1]
```diff
@@ -16,15 +17,62 @@
...
-                                      String expectedExceptionMessage) {
+                                      String expectedExceptionMessage,
+                                      Class[] parameters) {
         this.classes = classes;
         this.expectedTypeOfException = expectedTypeOfException;
         this.expectedExceptionMessage = expectedExceptionMessage;
+        this.parameters = parameters;
+    }
+
+    public interface CheckAble {
```
Comment (artem-zinnatullin) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java]: Maybe just `Checkable`?

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 2]
```diff
@@ -32,20 +80,14 @@ private PrivateConstructorChecker(List<Class> classes,
 
         private String expectedExceptionMessage;
 
+        private Class[] parameters;
```
Comment (artem-zinnatullin) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java]: Maybe `expectedParameters`, will improve readability of the code below

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 3]
```diff
@@ -138,16 +180,26 @@ private void check(Class clazz) {
...
         }
 
+        final Class<?>[] parameterTypes = constructor.getParameterTypes();
+
+        if (parameterTypes.length > 0) {
+            if (parameters == null) {
+                throw new AssertionError(clazz + " has non-default constructor with some parameters");
+            } else {
+                if (!Arrays.equals(parameterTypes, parameters)) {
+                    throw new AssertionError(clazz + " parameters do not match");
```
Comment (artem-zinnatullin) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java]: let's output expected and actual parameters

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 4]
```diff
@@ -138,16 +180,26 @@ private void check(Class clazz) {
...
 
         if (!Modifier.isPrivate(constructor.getModifiers())) {
             throw new AssertionError("Constructor of " + clazz + " must be private");
         }
 
+        final Class<?>[] parameterTypes = constructor.getParameterTypes();
+
+        if (parameterTypes.length > 0) {
+            if (parameters == null) {
+                throw new AssertionError(clazz + " has non-default constructor with some parameters");
```
Comment (artem-zinnatullin) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java:193]: Maybe `Expected class with constructor  without parameters but found constructor with parameters = …`?

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java (hunk 3 above)
Comment (artem-zinnatullin) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java]: And probably change format to `Expected constructor with parameters = … but found constructor with parameters = …`

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java [hunk 5]
```diff
@@ -138,16 +180,26 @@ private void check(Class clazz) {
...
-            throw new AssertionError(clazz + " has non-default constructor with some parameters");
-        }
-
         constructor.setAccessible(true);
 
         if (!Modifier.isPrivate(constructor.getModifiers())) {
             throw new AssertionError("Constructor of " + clazz + " must be private");
         }
 
+        final Class<?>[] parameterTypes = constructor.getParameterTypes();
```
Comment (artem-zinnatullin) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java:189]: So, it'll work only for classes with one private constructor? What is your use case of this? Just curious :smile:

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java (hunk 4 above)
Comment (vanniktech) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java:193]: This we should not change. That's still the old behavior. The only way parameter is null when we are using `expectedException` or `expectedMessage`

File: checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java (hunk 5 above)
Comment (vanniktech) [inline checker/src/main/java/com/pushtorefresh/private_constructor_checker/PrivateConstructorChecker.java:189]: Yup for now only with one private constructor. I want to guarantee that there's only one private constructor which takes a string as a parameter, which will be called from factory methods.

**Comment by vanniktech** [comment]
@artem-zinnatullin I updated the code with your suggestions.

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 6]
```diff
@@ -249,4 +249,140 @@ public void shouldThrowExceptionIfOneOfClassesHasNonDefaultConstructor() {
...
+        builder.check();
+    }
+
+    @Test
+    public void builderShouldThrowExceptionIfNullWasPassedAsParameters() {
+        final PrivateConstructorChecker.ParametersCheckable builder = PrivateConstructorChecker
+                .forClass(ClassWithPrivateConstructor.class);
+
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("parameters can not be null");
```
Comment (nikitin-da) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: expectedParameters can not be null

**Comment by nikitin-da** [comment]
Great :+1:

**Comment by vanniktech** [comment]
@nikitin-da I'm right now on vacation and will work on it once I'm back

**Comment by nikitin-da** [comment]
@vanniktech No problem. Have a good rest =)

**Comment by artem-zinnatullin** [comment]
@vanniktech have a good vacation! JFYI build is not passing on Travis :) 

Just ping us in future and we'll take a look again.

**Comment by vanniktech** [comment]
@nikitin-da & @artem-zinnatullin Thanks and I'm back. Travis also does build again.

**Comment by nikitin-da** [comment]
Thanks! :+1:

**Comment by artem-zinnatullin** [comment]
:+1:

**Comment by vanniktech** [comment]
Thank you guys :+1:

**Comment by artem-zinnatullin** [comment]
@vanniktech thank you!

1.2.0 was released today and should be available in Maven Central

**Comment by vanniktech** [comment]
@artem-zinnatullin thanks for releasing 1.2.0

**Pull request merged**